private_transactions/src/lib.rs

```diff
@@ -0,0 +1,782 @@
+			config: config,
+			notify: RwLock::new(Vec::new()),
+			transactions_for_signing: Mutex::new(SigningStore::new()),
+			transactions_for_verification: Mutex::new(VerificationStore::new()),
+			client: client,
+			accounts: accounts,
+			encryptor: encryptor,
+		})
+	}
+
+	/// Adds an actor to be notified on certain events
+	pub fn add_notify(&self, target: Arc<ChainNotify>) {
+		self.notify.write().push(Arc::downgrade(&target));
+	}
+
+	fn notify<F>(&self, f: F) where F: Fn(&ChainNotify) {
+		for np in self.notify.read().iter() {
+			if let Some(n) = np.upgrade() {
+				f(&*n);
+			}
+		}
+	}
+
+	/// 1. Create private transaction from the signed transaction
+	/// 2. Executes private transaction
+	/// 3. Save it with state returned on prev step to the queue for signing
+	/// 4. Broadcast corresponding message to the chain
+	pub fn create_private_transaction(&self, signed_transaction: SignedTransaction) -> Result<Receipt, PrivateTransactionError> {
+		trace!("Creating private transaction from regular transaction: {:?}", signed_transaction);
+		if self.config.signer_account.is_none() {
+			trace!("Signing account not set");
+			return Err(PrivateTransactionError::SignerAccountNotSet.into());
+		}
+		let tx_hash = signed_transaction.hash();
+		match signed_transaction.action {
+			Action::Create => {
+				return Err(PrivateTransactionError::BadTransactonType.into());
+			}
+			Action::Call(contract) => {
+				let data = signed_transaction.rlp_bytes();
+				let encrypted_transaction = self.encrypt(&contract, &Self::iv_from_transaction(&signed_transaction), &data)?;
+				let private = PrivateTransaction {
+					encrypted: encrypted_transaction,
+					contract: contract,
+				};
+				let contract_nonce = self.get_contract_nonce(&contract, BlockId::Latest)?;
+				let private_state = self.execute_private_transaction(BlockId::Latest, &signed_transaction)?;
+				trace!("Private transaction created, encrypted transaction: {:?}, private state: {:?}", private, private_state);
+				let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+				trace!("Required validators: {:?}", contract_validators);
+				let private_state_hash = self.calculate_state_hash(&private_state, contract_nonce);
+				trace!("Hashed effective private state for sender: {:?}", private_state_hash);
+				self.transactions_for_signing.lock().add_transaction(private.hash(), signed_transaction, contract_validators, private_state, contract_nonce)?;
+				self.broadcast_private_transaction(private.rlp_bytes().into_vec());
+				Ok(Receipt {
+					hash: tx_hash,
+					contract_address: None,
+					status_code: 0,
+				})
+			}
+		}
+	}
+
+	/// Calculate hash from united private state and contract nonce
+	fn calculate_state_hash(&self, state: &Bytes, nonce: U256) -> H256 {
+		let state_hash = keccak(state);
+		let mut state_buf = [0u8; 64];
+		state_buf[..32].clone_from_slice(&state_hash);
+		state_buf[32..].clone_from_slice(&H256::from(nonce).0);
+		keccak(&state_buf.as_ref())
```

Comment:
`as_ref()` is redundant

Reply:
error[E0277]: the trait bound `[u8; 64]: std::convert::AsRef<[u8]>` is not satisfied
   --> private_transactions/src/lib.rs:238:3
    |
238 |   keccak(&state_buf)
    |   ^^^^^^ the trait `std::convert::AsRef<[u8]>` is not implemented for `[u8; 64]`
    |
    = help: the following implementations were found:
              <[T; <unevaluated[]>] as std::convert::AsRef<[T]>>
              <[T; <unevaluated[]>] as std::convert::AsRef<[T]>>
              <[T; <unevaluated[]>] as std::convert::AsRef<[T]>>
              <[T; <unevaluated[]>] as std::convert::AsRef<[T]>>
            and 30 others
    = note: required because of the requirements on the impl of `std::convert::AsRef<[u8]>` for `&[u8; 64]`
    = note: required by `hash::keccak`